### PR TITLE
add pca tsne umap features

### PR DESCRIPTION
**Pull request by jchpan**

### CODE REVIEW

File: components/bokeh/__init__.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    source = ColumnDataSource(input_df)
+
+    TOOLTIPS = [
+        ("index", "$index"),
+        ("(x,y)", "($x, $y)"),
+        ("feature", '@'+feature),
+    ]
+
+    n = len(input_df[feature].unique())
+    p = figure(width=800, height=400, tooltips=TOOLTIPS,x_axis_label=x_lab, y_axis_label=y_lab)
```
Comment (u8sand) [inline components/bokeh/__init__.py]: It would be good if the figure would autosize itself to fit within the flex container instead of having an explicit size

File: components/bokeh/index.tsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import React from 'react'
+import { MetaNode } from '@/spec/metanode'
+
+const Plot = React.lazy(() => import('./bokeh'))
```
Comment (u8sand) [inline components/bokeh/index.tsx]: Use
```
import dynamic from 'next/dynamic'

const Plot = dynamic(() => import('./bokeh'), { ssr: false })
```

Instead of React.lazy

File: components/pca_transformation/__init__.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  df = anndata_input.to_df()
+  df = df.transpose()
+
+  # PCA
+  pca = PCA()
+  transformed_pca = pca.fit_transform(df)
+
+  # leiden coloring
+  leiden_df = sc.AnnData(df,dtype=np.float32)
+  sc.pp.pca(leiden_df)
```
Comment (u8sand) [inline components/pca_transformation/__init__.py]: 1. It looks like you do pca twice..? Just remove the sklearn stuff, scanpy is doing pca anyway, df is already anndata before to convert it to pandas df (remove that) the result of both PCA and leidan are stored in the AnnData object (probably it's df.obs['pca']).
2. Leiden is not a part of PCA, you should change the name (or description) of this step if you're doing both.

File: components/log_normalization/__init__.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import anndata as ad
+import pandas as pd
+import scanpy as sc
+from maayanlab_bioinformatics.normalization import log2_normalize
```
Comment (u8sand) [inline components/log_normalization/__init__.py]: This entire package is overkill for this. Instead just define:

```
def log2_normalize(x):
  return np.log2(x + 1.)

```

File: components/tsne_transformation/__init__.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+def tsne_transformation(gene_count_matrix):
+  anndata_input = anndata_from_path(gene_count_matrix['url'])
+  df = anndata_input.to_df()
+  df = df.transpose()
+
+  # t-SNE
+  n_jobs=mp.cpu_count()
+  tsne = TSNE(n_jobs = n_jobs, n_components=3)
+  transformed_tsne = tsne.fit_transform(df)
```
Comment (u8sand) [inline components/tsne_transformation/__init__.py]: As above, scanpy can do tsne: https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.tsne.html

File: components/umap_transformation/__init__.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+from components.bokeh import interactive_circle_plot
+
+def umap_transformation(gene_count_matrix):
+  anndata_input = anndata_from_path(gene_count_matrix['url'])
+  df = anndata_input.to_df()
+  df = df.transpose()
+
+  # UMAP
+  umap_model = umap.UMAP(n_components=3)
+  transformed_umap = umap_model.fit_transform(df)
```
Comment (u8sand) [inline components/umap_transformation/__init__.py]: As above, scanpy can do UMAP: https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.umap.html

File: components/umap_transformation/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+MulticoreTSNE
```
Comment (u8sand) [inline components/umap_transformation/requirements.txt]: You don't need the TSNE package to do umap.

File: components/z_score_normalization/__init__.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import scanpy as sc
+from components.gene_count_matrix import gene_count_matrix, anndata_from_path
+from components.file import upsert_file
+
+def z_score_normalize_gene_count_matrix(m):
+  anndata_input = anndata_from_path(m['url'])
+
+  # z-score normalization
+  norm_data = anndata_input.to_df()
+  normalized_df = pd.DataFrame(zscore(norm_data, axis=0), index=norm_data.index, columns=norm_data.columns)
```
Comment (u8sand) [inline components/z_score_normalization/__init__.py]: Z Score normalization of a gene count matrix is often done the other way (axis=1) to how it is expected to be oriented.

File: components/z_score_normalization/__init__.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
+from scipy.stats import zscore
+import pandas as pd
+import scanpy as sc
```
Comment (u8sand) [inline components/z_score_normalization/__init__.py]: This is not used, it shouldn't be imported or a requirement

File: components/umap_transformation/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+scanpy
+numpy
+pandas
+multiprocessing
```
Comment (u8sand) [inline components/umap_transformation/requirements.txt]: What is this package? multiprocessing is in the standard library

File: components/tsne_transformation/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,6 @@
+MulticoreTSNE
+scanpy
+numpy
+pandas
+multiprocessing
```
Comment (u8sand) [inline components/tsne_transformation/requirements.txt]: What is this package for?

File: components/quantile_normalization/__init__.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
+import qnorm
+import scanpy as sc
```
Comment (u8sand) [inline components/quantile_normalization/__init__.py]: You don't use scanpy here

**Review by u8sand — CHANGES_REQUESTED** [review]

File: components/bokeh/__init__.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    input_df['legend'] = input_df[feature]
+    source = ColumnDataSource(input_df)
+
+    TOOLTIPS = [
+        ("index", "$index"),
+        ("(x,y)", "($x, $y)"),
+        ("feature", '@'+feature),
+    ]
+
+    n = len(input_df[feature].unique())
```
Comment (u8sand) [inline components/bokeh/__init__.py]: Use `n = input_df[feature].nunique()`

**Comment by u8sand** [comment]
Please review the patches I made to this branch, addressing many of my comments and in order to get it deployed locally [juncheng.patch.tar.gz](https://github.com/nih-cfde/playbook-partnership/files/10026214/juncheng.tar.gz)

File: components/bokeh/__init__.py (hunk 13 above)
Comment (u8sand) [reply components/bokeh/__init__.py]: This is still unresolved

File: components/tsne_transformation/requirements.txt [hunk 14]
```diff
@@ -0,0 +1,5 @@
+bokeh<3
+numpy
+pandas
+scanpy
+sklearn
```
Comment (u8sand) [inline components/tsne_transformation/requirements.txt]: This should actually be scikit-learn

File: components/bokeh/__init__.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+from bokeh.transform import factor_cmap
+from bokeh.models import ColumnDataSource, Select, NumeralTickFormatter
+from bokeh.models.callbacks import CustomJS
+from bokeh.layouts import column, row
+import numpy as np
+import seaborn as sns
+
+def generate_colors(input_df, feature):
+    pal = sns.color_palette()
+    color = factor_cmap(feature, palette=pal.as_hex(), factors=np.array(input_df[feature].unique()))
```
Comment (u8sand) [inline components/bokeh/__init__.py]: I don't believe the outer `np.array(...)` here is necessary

File: components/bokeh/bokeh.tsx [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+    React.useEffect(() => {
+        if (!ref.current || !plot) return
+        const id = randid()
+        const div = document.createElement('div')
+        div.id = id
+        ref.current.appendChild(div)
+        Bokeh.embed.embed_item(plot, id)
+        return () => {
+            ref.current
```
Comment (u8sand) [inline components/bokeh/bokeh.tsx]: This exit part of the hook should be removing the previously created element.

`ref.current.removeChild(div)`

**Review by u8sand — CHANGES_REQUESTED** [review]
Just a few more minor changes left

File: components/bokeh/__init__.py (hunk 15 above)
Comment (jchpan) [reply components/bokeh/__init__.py]: We need some ways to remove dtype=object that is generated by unique().
I changed it to be input_df[feature].unique().tolist()

**Review by u8sand — APPROVED** [review]
